lessons/04-structured-output/README.md

```diff
@@ -0,0 +1,590 @@
+  }
+]
+```
+
+This is great, this output is something you can send to another service or application that can consume JSON data.
+
+**You:** If you say so.
+
+**Time beetle:** I do, I do.
+
+## Extracting data from prompts
+
+**Time beetle:** Imagine you're building a chatbot that helps users find information about trips to book. You want to extract key details from the user's input to provide relevant responses. Let's say you have a description like so: Type where you would like to go, when you want to travel, and your budget.
+
+Hopefully the user reads said instruction, what does a prompt look like that extracts the information? 
+
+> Prompt: Extract the following information from a users prompt: location, when, budget, currency, respond with the extracted data in JSON format, user prompt: I want to go to Greece in June, my budget is 800 euro.
+
+
+> Response:
+
+```json
+{
+  "location": "Greece",
+  "when": "June",
+  "budget": 800,
+  "currency": "euro"
+}
+```
+
+**Leonardo:** He doesn't stop talking does he?
+
+**You:** Let's see if I can find the off switch.
+
+**Time beetle:** Let's try this with a similar prompt, let's vary the user input a bit so we can be sure the model is doing the right thing.
+
+> Prompt: Extract the following information from a users prompt: location, when, budget, currency, respond with the extracted data in JSON format, user prompt: Going to US, max 1200 dollars, January maybe.
+>
+> Response:
+
+```json
+{
+  "location": "US",
+  "when": "January",
+  "budget": 1200,
+  "currency": "dollars"
+}
+```
+
+What you can do as a developer is then create a template like so:
+
+```text
+Extract the following information from a users prompt: location, when, budget, currency, respond with the extracted data in JSON format, user prompt: {user_input}.
+```
+
+**You:** Found it, take that George!
+
+**Time beetle:** _Powering down_
+
+**Leonardo:** Peace and quite at last.
+
+**Time beetle:** _Initiating emergency startup_ Where was I again? Oh yes.
+
+**Time beetle:** For chat bots, this is a very powerful feature as it allows you to extract information from the user's input. What if the user might ask for all sorts of things? You can create a template where you describe skills, things that your chatbot can do and what parameters it needs, here's how you can do that:
+
+```text
+Here are some skills you have:
+
+skill: book_trip
+parameters: location, when, budget, currency
+
+skill: find_hotel
+parameters: location, check_in, check_out, budget
+
+skill: order_food
+parameters: cuisine, delivery_address, payment_method
+
+Figure out which skill to use based on a users prompt and extract the necessary parameters, respond with the extracted data in JSON format with columns skill, parameters, extracted_data
+
+{user_input}
+```
+
+Let's try it with the following user inputs, that is this is the value for {user_input}:
```

Comment:
Let's try it with the following user inputs. Here is the value for {user_input}: